lib/util.c

```diff
@@ -312,6 +312,58 @@ __msp_safe_free(void **ptr)
     }
 }
 
+bool
+doubles_almost_equal(double a, double b, double eps)
+{
+    return (fabs(a) < eps && fabs(b) < eps) || gsl_fcmp(a, b, eps) == 0;
+}
+
+/* Inputs:
+ *   - `lengths` is an array of the real number lengths of `num_intervals`
+ *      consecutive half-open intervals on the real number line start at zero,
+ *      i.e. given lengths L_i, the intervals are:
+ *          [0.0, L_0), [L_0, L_0 + L_1), [L_0 + L_1, L_0 + L_1 + L_2), ...
```

Comment:
Given that the return is zero if `x` is negative aren't the intervals actually
```suggestion
 *          [-INFINITY, L_0), [L_0, L_0 + L_1), [L_0 + L_1, L_0 + L_1 + L_2), ...
```
??

Reply:
You're right. Conceptually the motivation was having intervals that have lengths equal to the inputs. And then a negative x or an x that is past all these intervals are a special case.
It probably would be more clear to just doc N+1 intervals extending all the way from -INF to +INF corresponding to all the x that can be passed in.